src/util/storable_builder.rs

```diff
@@ -0,0 +1,71 @@
+use crate::crypto::chacha20poly1305::ChaCha20Poly1305;
+use crate::types::{EncryptionMetadata, PlaintextBlob, Storable};
+use ::prost::Message;
+use rand::rngs::ThreadRng;
+use rand::RngCore;
+use std::borrow::Borrow;
+use std::io;
+use std::io::{Error, ErrorKind};
+
+pub struct StorableBuilder {
+	data_encryption_key: [u8; 32],
+}
+
+const CHACHA20_CIPHER_NAME: &'static str = "ChaCha20Poly1305";
+
+impl StorableBuilder {
+	pub fn build(&self, input: Vec<u8>, version: i64) -> Storable {
+		let mut rng = ThreadRng::default();
+		let mut nonce = [0u8; 12];
+		rng.fill_bytes(&mut nonce[4..]);
+
+		let mut data_blob = PlaintextBlob { value: input, version }.encode_to_vec();
+
+		let mut cipher = ChaCha20Poly1305::new(&self.data_encryption_key, &nonce, &vec![]);
+		let mut tag = [0u8; 16];
+		cipher.encrypt_inplace(&mut data_blob, &mut tag);
+		Storable {
+			data: data_blob,
+			encryption_metadata: Option::from(EncryptionMetadata {
+				nonce: nonce.to_vec(),
+				tag: tag.to_vec(),
+				cipher_format: CHACHA20_CIPHER_NAME.to_string(),
+			}),
+		}
+	}
+
+	pub fn deconstruct(&self, mut storable: Storable) -> io::Result<(Vec<u8>, i64)> {
+		let encryption_metadata = storable.encryption_metadata.unwrap();
+		let mut cipher = ChaCha20Poly1305::new(&self.data_encryption_key, &encryption_metadata.nonce, &vec![]);
+		let input_output = storable.data.as_mut();
+
+		if cipher.decrypt_inplace(input_output, encryption_metadata.tag.borrow()) {
+			let data_blob = PlaintextBlob::decode(&*input_output).map_err(|e| Error::new(ErrorKind::InvalidData, e))?;
```

Comment:
More idiomatic as:

```rust
if cipher.decrypt_inplace(&mut storable.data, &encryption_metadata.tag) {
	let data_blob = PlaintextBlob::decode(&storable.data[..]).map_err(|e| Error::new(ErrorKind::InvalidData, e))?;
}
```

Reply:
Hmm? My example dropped `input_output` entirely. Seems to just obscure what you are operating on, IMO. But my larger point was around not needing to explicitly use `as_mut` and `borrow`. Use `&mut` and `&` whenever possible.